<a href="https://colab.research.google.com/github/gtoubian/cce/blob/main/4_3_Regression_Workshop_Gabriel_Toubian.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Gabriel Toubian** - gtoubian@gmail.com

#NOTE:
Please insert a text cell at the top of the Notebook with your Full Name in Bold and your email that you use for both Journey and Github

Also, add to the title of the notebook, an underscore as well as your name.

Ex. '4.2_Regression_Workshop_Usama'


#Warmup

Import pandas, numpy, scipy.stats, statsmodels.api, LinearRegression from sklearn & ols from statsmodels.formula.api.

In [101]:
import pandas as pd
import numpy as np
import scipy.stats as sc
import statsmodels.api as sm
from statsmodels.formula.api import ols
from sklearn.linear_model import LinearRegression

#Boston Regression

Using statsmodels and the `boston` dataset, make a regression model to predict house prices. Don't forget to add a constant (intercept) term. Note that statsmodels can take a `pd.DataFrame` as an input for `X`.

Report the $R^2$ and coefficients on each feature.

In [102]:
from sklearn.datasets import load_boston
import pandas as pd

Without Intercept

In [ ]:
boston = pd.DataFrame(load_boston().data,columns = load_boston().feature_names )
y = load_boston().target
c= boston
sm.OLS(y, c).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                      y   R-squared (uncentered):                   0.959
Model:                            OLS   Adj. R-squared (uncentered):              0.958
Method:                 Least Squares   F-statistic:                              891.3
Date:                Fri, 26 Mar 2021   Prob (F-statistic):                        0.00
Time:                        13:56:02   Log-Likelihood:                         -1523.8
No. Observations:                 506   AIC:                                      3074.
Df Residuals:                     493   BIC:                                      3128.
Df Model:                          13                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
CRIM          -0.0929      0.034     -2.699      0.007      -0.161      -0.025
ZN             0.0487      0.014      3.382      0.001       0.020       0.077
INDUS         -0.0041      0.064     -0.063      0.950      -0.131       0.123
CHAS           2.8540      0.904      3.157      0.002       1.078       4.630
NOX           -2.8684      3.359     -0.854      0.394      -9.468       3.731
RM             5.9281      0.309     19.178      0.000       5.321       6.535
AGE           -0.0073      0.014     -0.526      0.599      -0.034       0.020
DIS           -0.9685      0.196     -4.951      0.000      -1.353      -0.584
RAD            0.1712      0.067      2.564      0.011       0.040       0.302
TAX           -0.0094      0.004     -2.395      0.017      -0.017      -0.002
PTRATIO       -0.3922      0.110     -3.570      0.000      -0.608      -0.176
B              0.0149      0.003      5.528      0.000       0.010       0.020
LSTAT         -0.4163      0.051     -8.197      0.000      -0.516      -0.317
==============================================================================
Omnibus:                      204.082   Durbin-Watson:                   0.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1374.225
Skew:                           1.609   Prob(JB):                    3.90e-299
Kurtosis:                      10.404   Cond. No.                     8.50e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.5e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

With Intercept

In [ ]:
y = load_boston().target

X = sm.add_constant(boston)
sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     108.1
Date:                Fri, 26 Mar 2021   Prob (F-statistic):          6.72e-135
Time:                        13:56:04   Log-Likelihood:                -1498.8
No. Observations:                 506   AIC:                             3026.
Df Residuals:                     492   BIC:                             3085.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.4595      5.103      7.144      0.000      26.432      46.487
CRIM          -0.1080      0.033     -3.287      0.001      -0.173      -0.043
ZN             0.0464      0.014      3.382      0.001       0.019       0.073
INDUS          0.0206      0.061      0.334      0.738      -0.100       0.141
CHAS           2.6867      0.862      3.118      0.002       0.994       4.380
NOX          -17.7666      3.820     -4.651      0.000     -25.272     -10.262
RM             3.8099      0.418      9.116      0.000       2.989       4.631
AGE            0.0007      0.013      0.052      0.958      -0.025       0.027
DIS           -1.4756      0.199     -7.398      0.000      -1.867      -1.084
RAD            0.3060      0.066      4.613      0.000       0.176       0.436
TAX           -0.0123      0.004     -3.280      0.001      -0.020      -0.005
PTRATIO       -0.9527      0.131     -7.283      0.000      -1.210      -0.696
B              0.0093      0.003      3.467      0.001       0.004       0.015
LSTAT         -0.5248      0.051    -10.347      0.000      -0.624      -0.425
==============================================================================
Omnibus:                      178.041   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              783.126
Skew:                           1.521   Prob(JB):                    8.84e-171
Kurtosis:                       8.281   Cond. No.                     1.51e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.51e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
X = X.drop(columns=['INDUS', 'AGE'])
sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.735
Method:                 Least Squares   F-statistic:                     128.2
Date:                Fri, 26 Mar 2021   Prob (F-statistic):          5.54e-137
Time:                        13:56:07   Log-Likelihood:                -1498.9
No. Observations:                 506   AIC:                             3022.
Df Residuals:                     494   BIC:                             3072.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.3411      5.067      7.171      0.000      26.385      46.298
CRIM          -0.1084      0.033     -3.307      0.001      -0.173      -0.044
ZN             0.0458      0.014      3.390      0.001       0.019       0.072
CHAS           2.7187      0.854      3.183      0.002       1.040       4.397
NOX          -17.3760      3.535     -4.915      0.000     -24.322     -10.430
RM             3.8016      0.406      9.356      0.000       3.003       4.600
DIS           -1.4927      0.186     -8.037      0.000      -1.858      -1.128
RAD            0.2996      0.063      4.726      0.000       0.175       0.424
TAX           -0.0118      0.003     -3.493      0.001      -0.018      -0.005
PTRATIO       -0.9465      0.129     -7.334      0.000      -1.200      -0.693
B              0.0093      0.003      3.475      0.001       0.004       0.015
LSTAT         -0.5226      0.047    -11.019      0.000      -0.616      -0.429
==============================================================================
Omnibus:                      178.430   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              787.785
Skew:                           1.523   Prob(JB):                    8.60e-172
Kurtosis:                       8.300   Cond. No.                     1.47e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.47e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Use polynomial features to improve your regression model in `2.1`. You can use squared and cubic features. Try to find a model that minimizes the `AIC` or `BIC` of your output table.

In [ ]:
A = np.array(X)
print(A)

[[1.0000e+00 6.3200e-03 1.8000e+01 ... 1.5300e+01 3.9690e+02 4.9800e+00]
 [1.0000e+00 2.7310e-02 0.0000e+00 ... 1.7800e+01 3.9690e+02 9.1400e+00]
 [1.0000e+00 2.7290e-02 0.0000e+00 ... 1.7800e+01 3.9283e+02 4.0300e+00]
 ...
 [1.0000e+00 6.0760e-02 0.0000e+00 ... 2.1000e+01 3.9690e+02 5.6400e+00]
 [1.0000e+00 1.0959e-01 0.0000e+00 ... 2.1000e+01 3.9345e+02 6.4800e+00]
 [1.0000e+00 4.7410e-02 0.0000e+00 ... 2.1000e+01 3.9690e+02 7.8800e+00]]


In [ ]:
x = sm.add_constant(A)
A = np.concatenate([x, (x[:,6] ** 3)[:, np.newaxis]], axis=1)
print(A)

[[1.00000000e+00 6.32000000e-03 1.80000000e+01 ... 3.96900000e+02
  4.98000000e+00 6.84179290e+01]
 [1.00000000e+00 2.73100000e-02 0.00000000e+00 ... 3.96900000e+02
  9.14000000e+00 1.22548701e+02]
 [1.00000000e+00 2.72900000e-02 0.00000000e+00 ... 3.92830000e+02
  4.03000000e+00 1.22548701e+02]
 ...
 [1.00000000e+00 6.07600000e-02 0.00000000e+00 ... 3.96900000e+02
  5.64000000e+00 1.01830369e+01]
 [1.00000000e+00 1.09590000e-01 0.00000000e+00 ... 3.93450000e+02
  6.48000000e+00 1.36330777e+01]
 [1.00000000e+00 4.74100000e-02 0.00000000e+00 ... 3.96900000e+02
  7.88000000e+00 1.57189376e+01]]


In [ ]:
A = pd.DataFrame(A)

In [ ]:
A.columns = ['const', 'CRIM', 'ZN', 'CHAS', 'NOX', 'RM', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'NEW']

In [ ]:
A

const     CRIM    ZN  CHAS  ...  PTRATIO       B  LSTAT         NEW
0      1.0  0.00632  18.0   0.0  ...     15.3  396.90   4.98   68.417929
1      1.0  0.02731   0.0   0.0  ...     17.8  396.90   9.14  122.548701
2      1.0  0.02729   0.0   0.0  ...     17.8  392.83   4.03  122.548701
3      1.0  0.03237   0.0   0.0  ...     18.7  394.63   2.94  222.787480
4      1.0  0.06905   0.0   0.0  ...     18.7  396.90   5.33  222.787480
..     ...      ...   ...   ...  ...      ...     ...    ...         ...
501    1.0  0.06263   0.0   0.0  ...     21.0  391.99   9.67   15.227175
502    1.0  0.04527   0.0   0.0  ...     21.0  396.90   9.08   11.969701
503    1.0  0.06076   0.0   0.0  ...     21.0  396.90   5.64   10.183037
504    1.0  0.10959   0.0   0.0  ...     21.0  393.45   6.48   13.633078
505    1.0  0.04741   0.0   0.0  ...     21.0  396.90   7.88   15.718938

[506 rows x 13 columns]

In [ ]:
sm.OLS(y, A).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.750
Model:                            OLS   Adj. R-squared:                  0.744
Method:                 Least Squares   F-statistic:                     123.3
Date:                Fri, 26 Mar 2021   Prob (F-statistic):          6.56e-140
Time:                        13:56:15   Log-Likelihood:                -1489.4
No. Observations:                 506   AIC:                             3005.
Df Residuals:                     493   BIC:                             3060.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         45.3890      5.397      8.411      0.000      34.786      55.992
CRIM          -0.1262      0.032     -3.887      0.000      -0.190      -0.062
ZN             0.0311      0.014      2.269      0.024       0.004       0.058
CHAS           2.6999      0.839      3.217      0.001       1.051       4.349
NOX          -23.5572      3.753     -6.276      0.000     -30.932     -16.183
RM             3.7789      0.399      9.466      0.000       2.995       4.563
DIS           -2.7115      0.335     -8.100      0.000      -3.369      -2.054
RAD            0.3357      0.063      5.342      0.000       0.212       0.459
TAX           -0.0138      0.003     -4.123      0.000      -0.020      -0.007
PTRATIO       -1.0246      0.128     -8.001      0.000      -1.276      -0.773
B              0.0092      0.003      3.517      0.000       0.004       0.014
LSTAT         -0.5278      0.047    -11.324      0.000      -0.619      -0.436
NEW            0.0121      0.003      4.343      0.000       0.007       0.018
==============================================================================
Omnibus:                      150.684   Durbin-Watson:                   1.107
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              576.550
Skew:                           1.313   Prob(JB):                    6.36e-126
Kurtosis:                       7.522   Cond. No.                     1.65e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.65e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Note AIC and BIC are lower now

Ajusted R-Squared is 0.744

#Regressive Tennis T-Test

Use a linear regression and statsmodels to run a t-test on whether Federer scores more points than his opponents in the `tennis.csv` dataset.


Also answer the following: should your regression include a constant term? Why or why not? How would it change the interpretation of your coefficient and p-value?

In [ ]:
from google.colab import files
data = files.upload()

Saving tennis.csv to tennis (2).csv


In [ ]:
import io
tennis = pd.read_csv(io.BytesIO(data['tennis.csv']))

In [ ]:
tennis = tennis.dropna()

In [ ]:
score = np.array(tennis['win'])
sc.ttest_1samp(score, 0.5)

Ttest_1sampResult(statistic=26.245156533243442, pvalue=2.6923903460915165e-116)

In [ ]:
def one_sample_one_tailed(sample_data, popmean, alpha=0.05, alternative='greater'):
    t, p = sc.ttest_1samp(sample_data, popmean)
    print ('t:',t)
    print ('p:',p)
    if alternative == 'greater' and (p/2 < alpha) and t > 0:
        print ('Reject Null Hypothesis for greater-than test')
    if alternative == 'less' and (p/2 < alpha) and t < 0:
        print ('Reject Null Hypothesis for less-than test')

In [ ]:
one_sample_one_tailed(score, 0.5)

t: 26.245156533243442
p: 2.6923903460915165e-116
Reject Null Hypothesis for greater-than test


p-value is way below 0.05. We reject the null hypothesis. Federer wins more than he loses, so he does score more points than his opponent.

In [ ]:
tennis1 = tennis
tennis1 = tennis1.drop(columns = ['win'])

In [ ]:
tennis1 = tennis1.dropna()

In [ ]:
tennis['win'] = tennis['win'].astype(int)

In [ ]:
tenniswin = tennis['win']

In [ ]:
tennis1

year  ... player2 total points total
0     1998  ...                      103.0
1     1998  ...                      106.0
2     1998  ...                      148.0
3     1998  ...                      130.0
5     1998  ...                      136.0
...    ...  ...                        ...
1173  2012  ...                      164.0
1174  2012  ...                      276.0
1175  2012  ...                       86.0
1176  2012  ...                       99.0
1177  2012  ...                      173.0

[1016 rows x 70 columns]

In [ ]:
tennis1.drop(tennis1.columns[[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,44,45]], axis = 1, inplace = True)

In [ ]:
tennis1.head()

year  ...  player2 total points total
0  1998  ...                       103.0
1  1998  ...                       106.0
2  1998  ...                       148.0
3  1998  ...                       130.0
5  1998  ...                       136.0

[5 rows x 48 columns]

In [ ]:
#without constant
e = tenniswin
d = tennis1

sm.OLS(e, d).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                    win   R-squared (uncentered):                   0.937
Model:                            OLS   Adj. R-squared (uncentered):              0.936
Method:                 Least Squares   F-statistic:                              785.2
Date:                Fri, 26 Mar 2021   Prob (F-statistic):                        0.00
Time:                        13:57:01   Log-Likelihood:                          67.771
No. Observations:                1016   AIC:                                     -97.54
Df Residuals:                     997   BIC:                                     -3.992
Df Model:                          19                                                  
Covariance Type:            nonrobust                                                  
=========================================================================================================
                                            coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------------
year                                      0.0003   1.29e-05     24.924      0.000       0.000       0.000
player1 double faults                     0.0066      0.005      1.358      0.175      -0.003       0.016
player1 1st serves in                     0.0011      0.000      2.341      0.019       0.000       0.002
player1 1st serves total                  0.0009      0.000      2.481      0.013       0.000       0.002
player1 1st serve points won             -0.0008      0.001     -0.550      0.583      -0.004       0.002
player1 1st serve points total            0.0011      0.000      2.341      0.019       0.000       0.002
player1 2nd serve points won              0.0047      0.001      4.272      0.000       0.003       0.007
player1 2nd serve points total           -0.0002      0.000     -0.566      0.571      -0.001       0.000
player1 break points won                  0.0278      0.006      4.897      0.000       0.017       0.039
player1 break points total               -0.0126      0.003     -4.748      0.000      -0.018      -0.007
player1 service games played              0.0021      0.007      0.299      0.765      -0.012       0.016
player1 1st serve return points won       0.0011      0.001      0.875      0.382      -0.001       0.004
player1 1st serve return points total     0.0006      0.000      1.390      0.165      -0.000       0.001
player1 2nd serve return points won       0.0015      0.001      1.478      0.140      -0.000       0.003
player1 2nd serve return points total     0.0005      0.000      1.933      0.054   -7.38e-06       0.001
player1 break points converted won        0.0007      0.008      0.096      0.923      -0.014       0.016
player1 break points converted total     -0.0004      0.002     -0.146      0.884      -0.005       0.004
player1 return games played              -0.0124      0.007     -1.776      0.076      -0.026       0.001
player1 total service points won          0.0039      0.001      2.717      0.007       0.001       0.007
player1 total service points total        0.0009      0.000      2.481      0.013       0.000       0.002
player1 total return points won           0.0026      0.001      1.984      0.047     2.9e-05       0.005
player1 total return points total         0.0011      0.001      2.114      0.035    7.92e-05       0.002
player1 total points won                  0.0065      0.002      3.147      0.002       0.002       0.011
player1 total points total                0.0020      0.001      2.512      0.012       0.000       0.004
player2 aces                              0.0001      0.002      0.050      0.960      -0.005       0.005
player2 double f

In [ ]:
#with constant
f = sm.add_constant(tennis1)
sm.OLS(e, f).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    win   R-squared:                       0.656
Model:                            OLS   Adj. R-squared:                  0.649
Method:                 Least Squares   F-statistic:                     99.95
Date:                Fri, 26 Mar 2021   Prob (F-statistic):          3.61e-215
Time:                        13:57:03   Log-Likelihood:                 67.777
No. Observations:                1016   AIC:                            -95.55
Df Residuals:                     996   BIC:                             2.919
Df Model:                          19                                         
Covariance Type:            nonrobust                                         
=========================================================================================================
                                            coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------------------
const                                    -0.4849      4.290     -0.113      0.910      -8.903       7.933
year                                      0.0006      0.002      0.264      0.792      -0.004       0.005
player1 double faults                     0.0067      0.005      1.340      0.180      -0.003       0.017
player1 1st serves in                     0.0011      0.000      2.341      0.019       0.000       0.002
player1 1st serves total                  0.0009      0.000      2.481      0.013       0.000       0.002
player1 1st serve points won             -0.0008      0.001     -0.556      0.578      -0.004       0.002
player1 1st serve points total            0.0011      0.000      2.341      0.019       0.000       0.002
player1 2nd serve points won              0.0047      0.001      4.269      0.000       0.003       0.007
player1 2nd serve points total           -0.0002      0.000     -0.566      0.571      -0.001       0.000
player1 break points won                  0.0278      0.006      4.895      0.000       0.017       0.039
player1 break points total               -0.0126      0.003     -4.747      0.000      -0.018      -0.007
player1 service games played              0.0020      0.007      0.291      0.771      -0.012       0.016
player1 1st serve return points won       0.0011      0.001      0.875      0.382      -0.001       0.004
player1 1st serve return points total     0.0006      0.000      1.385      0.166      -0.000       0.001
player1 2nd serve return points won       0.0015      0.001      1.478      0.140      -0.000       0.003
player1 2nd serve return points total     0.0005      0.000      1.935      0.053   -6.99e-06       0.001
player1 break points converted won        0.0007      0.008      0.093      0.926      -0.014       0.016
player1 break points converted total     -0.0004      0.002     -0.152      0.879      -0.005       0.004
player1 return games played              -0.0124      0.007     -1.771      0.077      -0.026       0.001
player1 total service points won          0.0039      0.001      2.712      0.007       0.001       0.007
player1 total service points total        0.0009      0.000      2.481      0.013       0.000       0.002
player1 total return points won           0.0026      0.001      1.984      0.047    2.89e-05       0.005
player1 total return points total         0.0011      0.001      2.115      0.035    7.95e-05       0.002
player1 total points won                  0.0065      0.002      3.144      0.002       0.002       0.011
player1 total points total                0.0020      0.001      2.512      0.012       0.000       0.004
player2 aces                              0.0001      0.002      0.042      0.966      -0.005       0.005
player2 d

Adding a constant does seem to have an effect on the coefficient and the p-value. I can generally see the p-value increasing when the constant is added, which we would like to avoid, so I would advise not to add a constant in this case

#College Admissions

Using the `college.csv` dataset, answer the following:

1. Do private schools see more admissions overall? Hint: use a binary explanatory variable for `Private`

2. Is the relation between `Top10perc` and `Top25perc` best fit using a model with only one variable, or one variable and a polynomial of degree 2? Is a constant term useful? How would you select for the best of these model specifications?

In [ ]:
from google.colab import files
data = files.upload()

Saving college.csv to college.csv


In [ ]:
college = pd.read_csv(io.BytesIO(data['college.csv']))

In [ ]:
college

Unnamed: 0 Private  ...  Expend  Grad.Rate
0      Abilene Christian University     Yes  ...    7041         60
1                Adelphi University     Yes  ...   10527         56
2                    Adrian College     Yes  ...    8735         54
3               Agnes Scott College     Yes  ...   19016         59
4         Alaska Pacific University     Yes  ...   10922         15
..                              ...     ...  ...     ...        ...
772         Worcester State College      No  ...    4469         40
773               Xavier University     Yes  ...    9189         83
774  Xavier University of Louisiana     Yes  ...    8323         49
775                 Yale University     Yes  ...   40386         99
776    York College of Pennsylvania     Yes  ...    4509         99

[777 rows x 19 columns]

In [ ]:
college1 = college[['Private', 'Accept']]
college1.head()

Private  Accept
0     Yes    1232
1     Yes    1924
2     Yes    1097
3     Yes     349
4     Yes     146

In [ ]:
Y= np.array(college1[college1['Private']=='Yes']['Accept'])
N= np.array(college1[college1['Private']=='No']['Accept'])

sc.f_oneway(Y,N)

F_onewayResult(statistic=226.11663957678556, pvalue=4.9536619983091194e-45)

The p-value is below 0.05, which indicates that we can reject the null hypothesis. Private schools do seem to have a higher rate of admission.

In [ ]:
m = college['Top10perc']
n = college['Top25perc']

**one variable**

In [ ]:
fit1 = sm.OLS(n, m).fit().summary()
fit1

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              Top25perc   R-squared (uncentered):                   0.911
Model:                            OLS   Adj. R-squared (uncentered):              0.911
Method:                 Least Squares   F-statistic:                              7971.
Date:                Fri, 26 Mar 2021   Prob (F-statistic):                        0.00
Time:                        13:57:39   Log-Likelihood:                         -3332.4
No. Observations:                 777   AIC:                                      6667.
Df Residuals:                     776   BIC:                                      6671.
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Top10perc      1.7275      0.019     89.281      0.000       1.690       1.766
==============================================================================
Omnibus:                      290.739   Durbin-Watson:                   1.499
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1310.349
Skew:                          -1.681   Prob(JB):                    2.89e-285
Kurtosis:                       8.401   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**One variable with constant**

In [ ]:
p = sm.add_constant(m)

In [ ]:
fit1constant = sm.OLS(n, p).fit().summary()
fit1constant

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              Top25perc   R-squared:                       0.796
Model:                            OLS   Adj. R-squared:                  0.795
Method:                 Least Squares   F-statistic:                     3018.
Date:                Fri, 26 Mar 2021   Prob (F-statistic):          1.87e-269
Time:                        13:57:48   Log-Likelihood:                -2805.2
No. Observations:                 777   AIC:                             5614.
Df Residuals:                     775   BIC:                             5624.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         28.1984      0.596     47.281      0.000      27.028      29.369
Top10perc      1.0014      0.018     54.933      0.000       0.966       1.037
==============================================================================
Omnibus:                       45.943   Durbin-Watson:                   1.920
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               85.462
Skew:                           0.401   Prob(JB):                     2.77e-19
Kurtosis:                       4.413   Cond. No.                         60.8
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**one variable and polynomial of degree 2**

In [ ]:
college['squared']=college['Top10perc']**2

In [ ]:
q = college[['Top10perc', 'squared']]

In [ ]:
fits = sm.OLS(n, q).fit().summary()
fits

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:              Top25perc   R-squared (uncentered):                   0.975
Model:                            OLS   Adj. R-squared (uncentered):              0.975
Method:                 Least Squares   F-statistic:                          1.515e+04
Date:                Fri, 26 Mar 2021   Prob (F-statistic):                        0.00
Time:                        13:58:01   Log-Likelihood:                         -2839.4
No. Observations:                 777   AIC:                                      5683.
Df Residuals:                     775   BIC:                                      5692.
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Top10perc      2.7198      0.025    110.830      0.000       2.672       2.768
squared       -0.0195      0.000    -44.515      0.000      -0.020      -0.019
==============================================================================
Omnibus:                      129.648   Durbin-Watson:                   1.927
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              287.613
Skew:                           0.919   Prob(JB):                     3.51e-63
Kurtosis:                       5.347   Cond. No.                         134.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**one variable and polynomial of degree 2 with constant**

In [ ]:
r = sm.add_constant(q)

In [ ]:
fitswithconstant = sm.OLS(n, r).fit().summary()
fitswithconstant

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              Top25perc   R-squared:                       0.852
Model:                            OLS   Adj. R-squared:                  0.851
Method:                 Least Squares   F-statistic:                     2220.
Date:                Fri, 26 Mar 2021   Prob (F-statistic):          2.50e-321
Time:                        13:58:38   Log-Likelihood:                -2681.0
No. Observations:                 777   AIC:                             5368.
Df Residuals:                     774   BIC:                             5382.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         16.6786      0.845     19.738      0.000      15.020      18.337
Top10perc      1.8140      0.050     36.229      0.000       1.716       1.912
squared       -0.0102      0.001    -17.072      0.000      -0.011      -0.009
==============================================================================
Omnibus:                      177.730   Durbin-Watson:                   1.960
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              570.823
Skew:                           1.089   Prob(JB):                    1.12e-124
Kurtosis:                       6.590   Cond. No.                     5.65e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.65e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**one variable and polynomial of degree 2 (but with no constant) has the highest "adjusted R-Squared", which indicates that it is most likely the best fit. I would also look at p-value as a major factor for feature selection, but they are all 0 (so equal)**

# Auto Multiple Regression

Using the `auto.csv` dataset, perform a multiple linear regression with `mpg` as the response and all other variables except name in `auto` as the predictors. Comment on the output:

i. Is there a relationship between the predictors and the response?

ii. Which predictors appear to have a statistically significant relationship to the response?

iii. What does the coefficient for the year variable suggest?

In [126]:
from google.colab import files
data = files.upload()

Saving auto.csv to auto (3).csv


In [127]:
auto = pd.read_csv("auto.csv", delim_whitespace = True)

In [128]:
auto=auto.dropna()

In [129]:
auto = auto.drop(auto[auto.horsepower == "?"].index)

In [130]:
auto.head()

mpg  cylinders  displacement  ...  year  origin                       name
1  18.0        8.0         307.0  ...  70.0     1.0  chevrolet chevelle malibu
2  15.0        8.0         350.0  ...  70.0     1.0          buick skylark 320
3  18.0        8.0         318.0  ...  70.0     1.0         plymouth satellite
4  16.0        8.0         304.0  ...  70.0     1.0              amc rebel sst
5  17.0        8.0         302.0  ...  70.0     1.0                ford torino

[5 rows x 9 columns]

In [133]:
auto.drop('name', inplace=True, axis=1)

In [134]:
auto.head()

mpg  cylinders  displacement horsepower  weight  acceleration  year  origin
1  18.0        8.0         307.0      130.0  3504.0          12.0  70.0     1.0
2  15.0        8.0         350.0      165.0  3693.0          11.5  70.0     1.0
3  18.0        8.0         318.0      150.0  3436.0          11.0  70.0     1.0
4  16.0        8.0         304.0      150.0  3433.0          12.0  70.0     1.0
5  17.0        8.0         302.0      140.0  3449.0          10.5  70.0     1.0

In [135]:
mpg = auto['mpg']

In [136]:
auto.drop('mpg', inplace=True, axis=1)

In [137]:
auto.head()

cylinders  displacement horsepower  weight  acceleration  year  origin
1        8.0         307.0      130.0  3504.0          12.0  70.0     1.0
2        8.0         350.0      165.0  3693.0          11.5  70.0     1.0
3        8.0         318.0      150.0  3436.0          11.0  70.0     1.0
4        8.0         304.0      150.0  3433.0          12.0  70.0     1.0
5        8.0         302.0      140.0  3449.0          10.5  70.0     1.0

In [142]:
mpg = mpg.astype(float)

In [144]:
auto = auto.astype(float)

In [145]:
t=auto
u=mpg

sm.OLS(u, t).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                    mpg   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.981
Method:                 Least Squares   F-statistic:                              2933.
Date:                Fri, 26 Mar 2021   Prob (F-statistic):                        0.00
Time:                        14:28:28   Log-Likelihood:                         -1030.4
No. Observations:                 392   AIC:                                      2075.
Df Residuals:                     385   BIC:                                      2103.
Df Model:                           7                                                  
Covariance Type:            nonrobust                                                  
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
cylinders       -0.7055      0.323     -2.181      0.030      -1.341      -0.070
displacement     0.0219      0.008      2.871      0.004       0.007       0.037
horsepower      -0.0396      0.013     -3.147      0.002      -0.064      -0.015
weight          -0.0060      0.001     -9.258      0.000      -0.007      -0.005
acceleration    -0.0796      0.090     -0.881      0.379      -0.257       0.098
year             0.5836      0.024     24.150      0.000       0.536       0.631
origin           1.3100      0.281      4.663      0.000       0.758       1.862
==============================================================================
Omnibus:                       46.837   Durbin-Watson:                   1.283
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               90.015
Skew:                           0.681   Prob(JB):                     2.84e-20
Kurtosis:                       4.912   Cond. No.                     6.00e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large,  6e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [ ]:
#i. Is there a relationship between the predictors and the response?